### PR TITLE
Add touch sleep wakeup API

### LINKED ISSUE #7431: ESP32-S3 touch wake up from deep sleep problem
### Board

ESP32-S3

### Device Description

Custom PCB, based on ESP32-S3FN8 

### Hardware Configuration

GPIO11 connected to touch pad. 

### Version

latest master (checkout manually)

### IDE Name

Arduino IDE

### Operating System

MacOS

### Flash frequency

40Mhz

### PSRAM enabled

no

### Upload speed

115200

### Description

I’m trying to use touch wake up from deep sleep using an ESP32-S3 in the Arduino framework. I have a custom pcb with a touch-pad connected to GPIO11 → T11

When I run the TouchRead example code I clearly get good results: when the touch pad is not touch I get readings of ±22920 and when I touch it goes to ±90000:

```tsx
12:13:25.552 -> 22921
12:13:26.527 -> 79766 // touched
12:13:27.535 -> 89048 
12:13:28.537 -> 90099
12:13:29.539 -> 90792
12:13:30.525 -> 22917 // released
12:13:31.552 -> 22920
12:13:32.554 -> 22924
12:13:33.541 -> 22926
```

When I use the touch wake up example from DeepSleep I don’t get good results. The ESP32 goes to sleep but wakes up instantly. I tried to set the threshold at different values: 1, 60000, 100000 etc. But regardless the ESP32 keep on having the same behaviour. Below you can see the behaviour. In the timestamp you can see it wakes up instantly, while it’s not being touched. 

```tsx
12:21:17.702 -> Wakeup was not caused by deep sleep: 0
12:21:17.739 -> Going to sleep now
12:21:17.923 -> ESP-ROM:esp32s3-20210327
12:21:17.923 -> Build:Mar 27 2021
12:21:17.958 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:17.958 -> SPIWP:0xee
12:21:17.958 -> mode:DIO, clock div:1
12:21:17.958 -> load:0x3fce3808,len:0x43c
12:21:17.958 -> load:0x403c9700,len:0xbec
12:21:17.958 -> load:0x403cc700,len:0x2a3c
12:21:17.958 -> entry 0x403c98d8
12:21:18.978 -> Boot number: 2
12:21:18.978 -> Wakeup caused by touchpad
12:21:19.014 -> Going to sleep now
12:21:19.121 -> ESP-ROM:esp32s3-20210327
12:21:19.121 -> Build:Mar 27 2021
12:21:19.121 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:19.155 -> SPIWP:0xee
12:21:19.155 -> mode:DIO, clock div:1
12:21:19.155 -> load:0x3fce3808,len:0x43c
12:21:19.155 -> load:0x403c9700,len:0xbec
12:21:19.155 -> load:0x403cc700,len:0x2a3c
12:21:19.155 -> entry 0x403c98d8
12:21:20.187 -> Boot number: 3
12:21:20.187 -> Wakeup caused by touchpad
12:21:20.187 -> Going to sleep now
12:21:20.261 -> ESP-ROM:esp32s3-20210327
12:21:20.261 -> Build:Mar 27 2021
12:21:20.261 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:20.261 -> SPIWP:0xee
12:21:20.261 -> mode:DIO, clock div:1
12:21:20.261 -> load:0x3fce3808,len:0x43c
12:21:20.261 -> load:0x403c9700,len:0xbec
12:21:20.261 -> load:0x403cc700,len:0x2a3c
12:21:20.261 -> entry 0x403c98d8
12:21:21.286 -> Boot number: 4
12:21:21.286 -> Wakeup caused by touchpad
12:21:21.320 -> Going to sleep now
```

I attached the example code. 

I notice that the touch wake-up behaviour of the ESP32-S3 is different than previous modules. In previous module the touch value was going down when being touched, while with the S3 it’s going up. 

Is there a way to get touch Wake-up to work within the Arduino framework with the ESP32-S3?

### Sketch

```cpp
/*
Deep Sleep with Touch Wake Up
=====================================
This code displays how to use deep sleep with
a touch as a wake up source and how to store data in
RTC memory to use it over reboots

This code is under Public Domain License.

Author:
Pranav Cherukupalli <cherukupallip@gmail.com>
*/

#define Threshold 60000 /* Greater the value, more the sensitivity */

RTC_DATA_ATTR int bootCount = 0;
touch_pad_t touchPin;
/*
Method to print the reason by which ESP32
has been awaken from sleep
*/
void print_wakeup_reason(){
  esp_sleep_wakeup_cause_t wakeup_reason;

  wakeup_reason = esp_sleep_get_wakeup_cause();

  switch(wakeup_reason)
  {
    case ESP_SLEEP_WAKEUP_EXT0 : Serial.println("Wakeup caused by external signal using RTC_IO"); break;
    case ESP_SLEEP_WAKEUP_EXT1 : Serial.println("Wakeup caused by external signal using RTC_CNTL"); break;
    case ESP_SLEEP_WAKEUP_TIMER : Serial.println("Wakeup caused by timer"); break;
    case ESP_SLEEP_WAKEUP_TOUCHPAD : Serial.println("Wakeup caused by touchpad"); break;
    case ESP_SLEEP_WAKEUP_ULP : Serial.println("Wakeup caused by ULP program"); break;
    default : Serial.printf("Wakeup was not caused by deep sleep: %d\n",wakeup_reason); break;
  }
}

/*
Method to print the touchpad by which ESP32
has been awaken from sleep
*/
void print_wakeup_touchpad(){
  touchPin = esp_sleep_get_touchpad_wakeup_status();

  switch(touchPin)
  {
    case 0  : Serial.println("Touch detected on GPIO 4"); break;
    case 1  : Serial.println("Touch detected on GPIO 0"); break;
    case 2  : Serial.println("Touch detected on GPIO 2"); break;
    case 3  : Serial.println("Touch detected on GPIO 15"); break;
    case 4  : Serial.println("Touch detected on GPIO 13"); break;
    case 5  : Serial.println("Touch detected on GPIO 12"); break;
    case 6  : Serial.println("Touch detected on GPIO 14"); break;
    case 7  : Serial.println("Touch detected on GPIO 27"); break;
    case 8  : Serial.println("Touch detected on GPIO 33"); break;
    case 9  : Serial.println("Touch detected on GPIO 32"); break;
    default : Serial.println("Wakeup not by touchpad"); break;
  }
}

void callback(){
  //placeholder callback function
}

void setup(){
  Serial.begin(115200);
  delay(1000); //Take some time to open up the Serial Monitor

  //Increment boot number and print it every reboot
  ++bootCount;
  Serial.println("Boot number: " + String(bootCount));

  //Print the wakeup reason for ESP32 and touchpad too
  print_wakeup_reason();
  //print_wakeup_touchpad();

  //Setup interrupt on Touch Pad 3 (GPIO15)
  touchAttachInterrupt(T11, callback, Threshold);

  //Configure Touchpad as wakeup source
  esp_sleep_enable_touchpad_wakeup();

  //Go to sleep now
  Serial.println("Going to sleep now");
  esp_deep_sleep_start();
  Serial.println("This will never be printed");
}

void loop(){
  //This will never be reached
}
```

### Debug Message

```plain
12:21:17.702 -> Boot number: 1
12:21:17.702 -> Wakeup was not caused by deep sleep: 0
12:21:17.739 -> Going to sleep now
12:21:17.923 -> ESP-ROM:esp32s3-20210327
12:21:17.923 -> Build:Mar 27 2021
12:21:17.958 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:17.958 -> SPIWP:0xee
12:21:17.958 -> mode:DIO, clock div:1
12:21:17.958 -> load:0x3fce3808,len:0x43c
12:21:17.958 -> load:0x403c9700,len:0xbec
12:21:17.958 -> load:0x403cc700,len:0x2a3c
12:21:17.958 -> entry 0x403c98d8
12:21:18.978 -> Boot number: 2
12:21:18.978 -> Wakeup caused by touchpad
12:21:19.014 -> Going to sleep now
12:21:19.121 -> ESP-ROM:esp32s3-20210327
12:21:19.121 -> Build:Mar 27 2021
12:21:19.121 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:19.155 -> SPIWP:0xee
12:21:19.155 -> mode:DIO, clock div:1
12:21:19.155 -> load:0x3fce3808,len:0x43c
12:21:19.155 -> load:0x403c9700,len:0xbec
12:21:19.155 -> load:0x403cc700,len:0x2a3c
12:21:19.155 -> entry 0x403c98d8
12:21:20.187 -> Boot number: 3
12:21:20.187 -> Wakeup caused by touchpad
12:21:20.187 -> Going to sleep now
12:21:20.261 -> ESP-ROM:esp32s3-20210327
12:21:20.261 -> Build:Mar 27 2021
12:21:20.261 -> rst:0x5 (DSLEEP),boot:0x8 (SPI_FAST_FLASH_BOOT)
12:21:20.261 -> SPIWP:0xee
12:21:20.261 -> mode:DIO, clock div:1
12:21:20.261 -> load:0x3fce3808,len:0x43c
12:21:20.261 -> load:0x403c9700,len:0xbec
12:21:20.261 -> load:0x403cc700,len:0x2a3c
12:21:20.261 -> entry 0x403c98d8
12:21:21.286 -> Boot number: 4
12:21:21.286 -> Wakeup caused by touchpad
12:21:21.320 -> Going to sleep now
```

### Other Steps to Reproduce

Issue also occurs using the ESP32-S3 dev board

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by P-R-O-C-H-Y**
## Description of Change
As reported in the issue #7431 , the example `DeepSleep/TouchWakeUp.ino` is not working for ESP32-S2 and ESP32-S3. For these chips setting touch pad as wake up source needs different calls than ESP32. 
So this new function works with all chips.
`void touchSleepWakeUpEnable(uint8_t pin, touch_value_t threshold);`

I have updated the example to work using this new function and made it work on all chips + updated docs.

## Tests scenarios

Tested on ESP32, ESP32-S2 and ESP32-S3 with the updated `TouchWakeUp.ino` example.

## Related links
Closes #7431

### CODE REVIEW

**Review by me-no-dev — APPROVED** [review]

**Review by SuGlider — APPROVED** [review]

**Comment by SuGlider** [comment]
Very good, @P-R-O-C-H-Y !! I liked it!

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino [hunk 1]
```diff
@@ -11,7 +11,13 @@ Author:
...
 */
 
-#define Threshold 40 /* Greater the value, more the sensitivity */
+#if CONFIG_IDF_TARGET_ESP32
+  #define THRESHOLD 40      /* Greater the value, more the sensitivity */
+#elif CONFIG_IDF_TARGET_ESP32S2
+  #define THRESHOLD   30000 /* Lower the value, more the sensitivity */
+#else //CONFIG_IDF_TARGET_ESP32S3 + default for other chips (to be adjusted) */
+  #define THRESHOLD   80000 /* Lower the value, more the sensitivity */
+#endif
```
Comment (SuGlider) [inline libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:20]: I have tested it and it works better with this setup:

ESP32 -> Threshold = 40
S2/S3 -> Threshold = 1000

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino [hunk 2]
```diff
@@ -74,11 +87,8 @@ void setup(){
   print_wakeup_reason();
   print_wakeup_touchpad();
 
-  //Setup interrupt on Touch Pad 3 (GPIO15)
-  touchAttachInterrupt(T3, callback, Threshold);
-
-  //Configure Touchpad as wakeup source
-  esp_sleep_enable_touchpad_wakeup();
+  //Setup sleep wakeup on Touch Pad 3 (GPIO15 for ESP32) / (GPIO3 for ESP32-S2 and S3)
+  touchSleepWakeUpEnable(T3,THRESHOLD);
```
Comment (SuGlider) [inline libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:91]: I have tested it adding 2 touch pin to the sketch:

```touchSleepWakeUpEnable(T2,THRESHOLD);
touchSleepWakeUpEnable(T7,THRESHOLD);```

ESP32 wakes up with both and reports the right pin.
S2/S3 only wakes up with the last Touch Pad used in `touchSleepWakeUpEnable()`, in the example above, it means the the SoC only wakes up when touching GPIO7.

**Review by SuGlider — CHANGES_REQUESTED** [review]

**Comment by SuGlider** [comment]
@P-R-O-C-H-Y - It seems there is a problem with S2/S3 when the sketch uses more than 1 touchpad for waking up. It works fine with ESP32, but not with S2/S3.

**Comment by P-R-O-C-H-Y** [comment]
> @P-R-O-C-H-Y - It seems there is a problem with S2/S3 when the sketch uses more than 1 touchpad for waking up. It works fine with ESP32, but not with S2/S3. 
> 

@SuGlider I think S2/S3 can have only 1 touchpad for waking up. I have read about that in IDF docs. Maybe will be good if thats true to add a note in our docs too.

**Comment by P-R-O-C-H-Y** [comment]
@SuGlider Link to docs [link](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/touch_pad.html#_CPPv430touch_pad_sleep_channel_enable11touch_pad_tb)

**Comment by SuGlider** [comment]
> @SuGlider Link to docs [link](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/touch_pad.html#_CPPv430touch_pad_sleep_channel_enable11touch_pad_tb)

Correct, I see that. This is valid for the S2 and S3. Please add a note to the documentation. Thanks!

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino (hunk 1 above)
Comment (P-R-O-C-H-Y) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:20]: For ESP32 we do have threshold 40.
But for S2/S3 it cannot be that small. Normal reading is higher that this value.
For S2 if the pad is touched, reading from touchRead is above 45k. For S3 its above 95k

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino (hunk 2 above)
Comment (P-R-O-C-H-Y) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:91]: Do we need to add more pins for wakeup to the example?

**Comment by P-R-O-C-H-Y** [comment]
@SuGlider Docs updated with the note.

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino (hunk 1 above)
Comment (SuGlider) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:20]: I think that this is actually the delta, not the reading value itself. 
It worked better (faster when touching the S2/S3 pad) when I changed it to 1,000
You can try it to check it.

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino (hunk 2 above)
Comment (SuGlider) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:91]: Maybe... when it is an ESP32 only. Just to demonstrate it?! 
Also, maybe, add commentaries about the limitation for the S2/S3 in the sketch?
Just a few suggestions.

Comment (P-R-O-C-H-Y) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino]: Added both :) 
- Demontration for ESP32
- Comment in header about limitation

File: libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino (hunk 1 above)
Comment (P-R-O-C-H-Y) [reply libraries/ESP32/examples/DeepSleep/TouchWakeUp/TouchWakeUp.ino:21]: I have tested that a bit now, and managed to change it to 5000. That worked the best for me.
The 1000 was too much sensitive. I have attached the Dupont cable to the pin and it triggered even if I touched the rubber part of the cable, not the wire.
The 5000 triggers immediately I touch the wire :)

**Review by SuGlider — APPROVED** [review]

**Comment by rubenfrolic** [comment]
How do I get this to work in Arduino IDE? I tried to find a 2.0.5 dev version of the core. but can't find it. But need to update the core for this to work

**Comment by rubenfrolic** [comment]
Seems like it got it to work using the manual installation process: https://docs.espressif.com/projects/arduino-esp32/en/latest/installing.html

**Pull request merged**